contrib/scalar/scalar.c

```diff
@@ -19,55 +19,91 @@ static int is_unattended(void) {
 	return git_env_bool("Scalar_UNATTENDED", 0);
 }
 
+/**
+ * Remove the deepest subdirectory in the provided path string. Path must not
+ * include a trailing path separator. Returns 1 if parent directory found,
+ * otherwise 0.
+ */
+static int strbuf_parentdir(struct strbuf *buf)
+{
+	size_t len = buf->len;
+	size_t offset = offset_1st_component(buf->buf);
+	char *path_sep = find_last_dir_sep(buf->buf + offset);
+	strbuf_setlen(buf, path_sep ? path_sep - buf->buf : offset);
+
+	return buf->len < len;
+}
+
 static void setup_enlistment_directory(int argc, const char **argv,
 				       const char * const *usagestr,
-				       const struct option *options)
+				       const struct option *options,
+				       struct strbuf *enlistment_root)
 {
+	struct strbuf path = STRBUF_INIT;
+	char *root;
+	int enlistment_found = 0;
+
 	if (startup_info->have_repository)
 		BUG("gitdir already set up?!?");
 
 	if (argc > 1)
 		usage_with_options(usagestr, options);
 
+	/* find the worktree, determine its corresponding root */
 	if (argc == 1) {
-		char *src = xstrfmt("%s/src", argv[0]);
-		const char *dir = is_directory(src) ? src : argv[0];
+		strbuf_add_absolute_path(&path, argv[0]);
+	} else if(strbuf_getcwd(&path) < 0) {
+		die(_("need a working directory"));
+	}
 
-		if (chdir(dir) < 0)
-			die_errno(_("could not switch to '%s'"), dir);
+	strbuf_trim_trailing_dir_sep(&path);
+	do {
+		const size_t len = path.len;
 
-		free(src);
-	} else {
-		/* find the worktree, and ensure that it is named `src` */
-		struct strbuf path = STRBUF_INIT;
+		/* check if currently in enlistment root with src/ workdir */
+		strbuf_addstr(&path, "/src/.git");
+		if (is_git_directory(path.buf)) {
+			strbuf_strip_suffix(&path, "/.git");
 
-		if (strbuf_getcwd(&path) < 0)
-			die(_("need a working directory"));
-
-		for (;;) {
-			size_t len = path.len;
-
-			strbuf_addstr(&path, "/src/.git");
-			if (is_git_directory(path.buf)) {
-				strbuf_setlen(&path, len);
-				strbuf_addstr(&path, "/src");
-				if (chdir(path.buf) < 0)
-					die_errno(_("could not switch to '%s'"),
-						  path.buf);
-				strbuf_release(&path);
-				break;
-			}
+			if (enlistment_root)
+				strbuf_add(enlistment_root, path.buf, len);
 
-			while (len > 0 && !is_dir_sep(path.buf[--len]))
-				; /* keep looking for parent directory */
+			enlistment_found = 1;
+			break;
+		}
 
-			if (!len)
-				die(_("could not find enlistment root"));
+		/* reset to original path */
+		strbuf_setlen(&path, len);
 
+		/* check if currently in workdir */
+		strbuf_addstr(&path, "/.git");
+		if (is_git_directory(path.buf)) {
 			strbuf_setlen(&path, len);
+
+			if (enlistment_root) {
+				/*
+				* If the worktree's directory's name is `src`, the enlistment is the
+				* parent directory, otherwise it is identical to the worktree.
+				*/
```

Comment:
nit: spacing (probably got messed up with an auto-tabbing feature)
```suggestion
				/*
				 * If the worktree's directory's name is `src`, the enlistment is the
				 * parent directory, otherwise it is identical to the worktree.
				 */
```

Reply:
✅